kubeflow/hub/api/model_registry_client.py

```diff
@@ -0,0 +1,285 @@
+from __future__ import annotations
+
+from collections.abc import Callable, Coroutine, Iterator, Mapping
+import logging
+from typing import TYPE_CHECKING, Any, TypeVar
+
+from typing_extensions import Self
+
+if TYPE_CHECKING:
+    from model_registry import ModelRegistry
+    from model_registry.types import (
+        ModelArtifact,
+        ModelVersion,
+        RegisteredModel,
+        SupportedTypes,
+    )
+
+T = TypeVar("T")
+
+
+class ModelRegistryClient:
+    """Client for Kubeflow Model Registry operations.
+
+    Requires the model-registry package to be installed. Install it with:
+
+        pip install 'kubeflow[hub]'
+
+    """
+
+    def __init__(
+        self,
+        server_address: str,
+        port: int = 443,
+        *,
+        author: str,
+        is_secure: bool = True,
+        user_token: str | None = None,
+        user_token_envvar: str | None = None,
+        custom_ca: str | None = None,
+        custom_ca_envvar: str | None = None,
+        log_level: int = logging.WARNING,
+        async_runner: Callable[[Coroutine[Any, Any, T]], T] | None = None,
+    ):
+        """Initialize the ModelRegistryClient.
+
+        Args:
+            server_address: Server address.
+            port: Server port. Defaults to 443.
+
+        Keyword Args:
+            author: Name of the author.
+            is_secure: Whether to use a secure connection. Defaults to True.
+            user_token: The PEM-encoded user token as a string.
+            user_token_envvar: Environment variable to read the user token from.
+            custom_ca: Path to the PEM-encoded root certificates as a string.
+            custom_ca_envvar: Environment variable to read the custom CA from.
+            log_level: Log level. Defaults to logging.WARNING.
+            async_runner: A modular async scheduler Callable that takes in a coroutine.
+
+        Raises:
+            ImportError: If model-registry is not installed.
+        """
+        try:
+            from model_registry import ModelRegistry
+        except ImportError as e:
+            raise ImportError(
+                "model-registry is not installed. Install it with:\n\n"  # fmt: skip
+                "  pip install 'kubeflow[hub]'\n"
+            ) from e
```

Comment:
Note, the code currently allows the user to import `from kubeflow.hub import ModelRegistryClient` without error if `model-registry` is missing. It's not until `ModelRegistryClient()` is called that they'll get an error. I believe this makes sense.